### PR TITLE
Enable client turrets to track targets

### PR DESCRIPTION
**Pull request by JohnAFernandez**
This allows clients to user server data to move turrets according to the timing and target data on the server.  Uses a small packet to do so.

### CODE REVIEW

File: code/ai/aicode.cpp
Comment (BMagnu) [inline code/ai/aicode.cpp:11745]: Will this not set cur_angle each frame?
Intuitively, this should be set once when the packet arrives, and from then the client tracks the transmitted target itself based on the turret_enemy_objnum, no?
Or am I missing something?

**Review by BMagnu — COMMENTED** [review]
Looks sensible, but I have one question

File: code/ai/aicode.cpp
Comment (JohnAFernandez) [reply code/ai/aicode.cpp:11745]: Ah, good question! It will only set it once because, `Multi_Turret_Manager.get_latest_packet()` discards all packets except those that are not yet elapsed. Those that haven't elapsed are added to the `retained` vector that is made the value in the map for that net signature/subsystem pair.

File: code/network/multi_turret_manager.cpp
Comment (BMagnu) [inline code/network/multi_turret_manager.cpp:48]: this should really be moved, otherwise you'll be making unecessary copies of all elements
```suggestion
        it->second = std::move(retained);
```

File: code/network/multi_turret_manager.cpp
Comment (BMagnu) [inline code/network/multi_turret_manager.cpp:55]: If this is the case, none of the packets was yet available. This can also be formulated as "all packets will be available in the future".
In that case, why do you need the copy at all? should it->second not still be correct as is then?

**Review by BMagnu — CHANGES_REQUESTED** [review]
I see how it works now.
If that is the case, some small performance changes in your turret manager

**Review by BMagnu — APPROVED** [review]

File: code/network/multi_turret_manager.h
Comment (notimaginative) [inline code/network/multi_turret_manager.h:22]: `multi_get_network_object()` can return `nullptr`, so you should always check its return value before use. Particularly in this case since it would result in an invalid `target_objnum`. You should get the target object, verify it's valid, then set `target_objnum` accordingly (OBJ_INDEX or -1).

File: code/network/multimsgs.cpp
Comment (notimaginative) [inline code/network/multimsgs.cpp:9160]: The subsys count should be the first thing in the packet to make it easier to update. Move it first, get rid of the magic number for its position, and just use `HEADER_LENGTH` as the offset when updating.

And to avoid confusion it should really be set to zero by default instead of the max possible count;

File: code/network/multimsgs.cpp
Comment (notimaginative) [inline code/network/multimsgs.cpp:9199]: You should change that count > 0 here before sending and avoid wasting bandwidth on useless packets.

File: code/network/multimsgs.cpp
Comment (notimaginative) [inline code/network/multimsgs.cpp:9134]: It would probably be good to verify that `subsys_indexes` isn't empty before continuing with this. Though really, it shouldn't even get here if that is the case.

File: code/network/multi_turret_manager.h
Comment (notimaginative) [inline code/network/multi_turret_manager.h:91]: You should verify that `queued.second` actually has data before processing the packet. As `_packets_out` still contains all of the packets which have previously been processed, it's loaded with useless data which is just wasting time and bandwidth.

File: freespace2/freespace.cpp
Comment (notimaginative) [inline freespace2/freespace.cpp:1056]: This appears to be a cleanup function so it should really go in `game_level_close()` instead.

File: code/network/multimsgs.cpp
Comment (notimaginative) [inline code/network/multimsgs.cpp:9120]: If my math is right then you should be able to add about 170 turrets per packet. Best to just slave this to the max packet size rather than some magic number.

Inside the sending function set a const variable to `MAX_PACKET_SIZE - 20` (technically the 20 is a magic number too, but it's a small safety guard). Then add turrets until `packet_size` hits this variable or you've added all the turrets.

**Review by notimaginative — CHANGES_REQUESTED** [review]

File: code/network/multimsgs.cpp
Comment (JohnAFernandez) [reply code/network/multimsgs.cpp:9134]: Well, that data would be nonsense because it's attached to a net signature of 0 and nothing should be done with it. So this is a way for us to gracefully tell the manager to knock it off.  But it really shouldn't try to send info if the net signature is 0 in the first place. 

How about adding an assertion that `net_signature != 0` and keeping this as a valid way to fail in the quite unlikely event that the manager tries to send this packet in release?

File: code/network/multi_turret_manager.h
Comment (JohnAFernandez) [reply code/network/multi_turret_manager.h:91]: Yes, you're right that I should check if it's empty so that we don't end up sending an empty packet.

But part of the design of `send_turret_tracking_packet` is to return only the subsystems that there wasn't room for in the packet for the manager to reattempt later.  So it won't contain the packets already sent.  It will still have the ship in the unordered map, but the vector in most cases will be empty after the packet is sent.

File: code/network/multimsgs.cpp
Comment (JohnAFernandez) [reply code/network/multimsgs.cpp:9120]: If we had a lot of packets to write, I would suggest making a global `MAX_SAFE_PACKET_SIZE` define any packet could use.  But, this far along I don't see the point.

I think you're right about comparing it to max packet size, because the pack function for the angles is variable as well.

File: code/network/multimsgs.cpp
Comment (JohnAFernandez) [reply code/network/multimsgs.cpp:9134]: Wait, you meant that I should also check if the list is empty?

File: code/network/multimsgs.cpp
Comment (notimaginative) [reply code/network/multimsgs.cpp]: There are one or two other examples in the code where it's done this way (though I don't remember them off-hand). MAX_PACKET_SIZE is already the largest safe packet size though, the extra 20 bytes is so that you don't have to get super tricky with the size check.

So you can safely do `while (packet_size < MY_MAX_SIZE) {...}` and not have to worry about the exact size of the data being added. Each turret looks to add about 7 bytes so it should end the `while` loop with 13 bytes to spare. If that makes sense.

Comment (JohnAFernandez) [reply code/network/multimsgs.cpp]: I think I'm just going to keep it at 20 because in case anyone has to change this packet in the future, it would probably be good to have a good amount of padding.  I don't think it's going to be that often that a ship hits the limit.

**Review by notimaginative — APPROVED** [review]

**Comment by wookieejedi** [comment]
Just adding here that given this needs a multi bump, we'll probably wait until #4836 is approved to merge both.

**Pull request merged**
